flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutor.java

```diff
@@ -348,20 +353,41 @@ private void handleStartTaskExecutorServicesException(Exception e) throws Except
 		resourceManagerHeartbeatManager.stop();
 
 		try {
-			stopTaskExecutorServices();
+			stopServicesBeforeTasksCompletion();
 		} catch (Exception e) {
 			throwable = ExceptionUtils.firstOrSuppressed(e, throwable);
 		}
 
 		if (throwable != null) {
 			return FutureUtils.completedExceptionally(new FlinkException("Error while shutting the TaskExecutor down.", throwable));
 		} else {
-			log.info("Stopped TaskExecutor {}.", getAddress());
-			return CompletableFuture.completedFuture(null);
+			return FutureUtils.waitForAll(taskCompletionFutures).thenRun(() -> {
```

Comment:
If `throwable != null`, then I think we should still try to call `stopServicesAfterTasksCompletion`